source/lattice_sum.cpp

```diff
@@ -144,12 +144,18 @@ Complex3D structure_factor_spline(
         r[2] = m[2][0]*v[0] + m[2][1]*v[1] + m[2][2]*v[2];
         return r;
     };
-    // compute fractional coordinates
+    // compute fractional coordinates and make sure they lie in [0,1)
+    // the repeated operation is intentional to treat special cases
+    // where the initial frac_coords[i][j] is small and negative (e.g. -1e-16)
     std::vector<std::array<double,3>> frac_coords(xyz_coords.size());
     for (size_t i=0; i<frac_coords.size(); ++i) {
         frac_coords[i] = m_dot_v(ainv, xyz_coords[i]);
-        //for (size_t j=0; j<3; ++j)
-        //    frac_coords[i][j] -= std::floor(frac_coords[i][j]);
+        for (size_t j=0; j<3; ++j) {
+            frac_coords[i][j] -= std::floor(frac_coords[i][j]);
+        }
+        for (size_t j=0; j<3; ++j) {
```

Comment:
Would it make sense to do (roughly) `frac_coords = abs(frac_coords)` instead of repeating the operation?

Reply:
After chatting, we decided to look at how ASE does this.

Reply:
I've looked now and ASE does exactly what you propose in its `get_scaled_positions` routine. So let's stick with that.

But would you please delete the second loop and instead perform the operation twice in the first loop? Then move the comment about the duplication inside that loop and revise the comment slightly to explicitly say that, without it, values like -1e-20 are sometimes mapped to 1.0.